osutil/disks/mockdisk.go

```diff
@@ -0,0 +1,118 @@
+// without needing to mock the mount setup, sysfs, or udev commands just to test
+// high level logic.
+// DevNum must be a unique string per unique mocked disk, if only one disk is
+// being mocked it can be left empty.
+type MockDiskMapping struct {
+	FilesystemLabelToPartUUID map[string]string
+	DiskHasPartitions         bool
+	DevNum                    string
+}
+
+// FindMatchingPartitionUUID returns a matching PartitionUUID for the specified
+// label if it exists. Part of the Disk interface.
+func (d *MockDiskMapping) FindMatchingPartitionUUID(label string) (string, error) {
+	osutil.MustBeTestBinary("mock disks only to be used in tests")
+	if partuuid, ok := d.FilesystemLabelToPartUUID[label]; ok {
+		return partuuid, nil
+	}
+	fmt := "could not find label %q: %w"
+	return "", xerrors.Errorf(fmt, label, ErrFilesystemLabelNotFound)
+}
+
+// HasPartitions returns if the mock disk has partitions or not. Part of the
+// Disk interface.
+func (d *MockDiskMapping) HasPartitions() bool {
+	osutil.MustBeTestBinary("mock disks only to be used in tests")
+	return d.DiskHasPartitions
+}
+
+// MountPointIsFromDisk returns if the disk that the specified mount point comes
+// from is the same disk as the object. Part of the Disk interface.
+func (d *MockDiskMapping) MountPointIsFromDisk(mountpoint string, opts *Options) (bool, error) {
+	osutil.MustBeTestBinary("mock disks only to be used in tests")
+
+	// this is relying on the fact that DiskFromMountPoint should have been
+	// mocked for us to be using this mockDisk method anyways
+	otherDisk, err := DiskFromMountPoint(mountpoint, opts)
+	if err != nil {
+		return false, err
+	}
+
+	if otherDisk.Dev() == d.Dev() && otherDisk.HasPartitions() == d.HasPartitions() {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// Dev returns a unique representation of the mock disk, it is a hash of the
+// mock disk struct string representation. Part of the Disk interface.
+func (d *MockDiskMapping) Dev() string {
+	osutil.MustBeTestBinary("mock disks only to be used in tests")
+	return d.DevNum
+}
+
+// Mountpoint is a combination of a mountpoint location and whether that
+// mountpoint is a decrypted device. It is only used in identifying mount points
+// with MountPointIsFromDisk and DiskFromMountPoint with
+// MockMountPointDisksToPartionMapping.
+type Mountpoint struct {
+	Mountpoint        string
+	IsDecryptedDevice bool
+}
+
+// MockMountPointDisksToPartionMapping will mock DiskFromMountPoint such that
+// the specified mapping is returned/used. Specifically, keys in the provided
+// map are mountpoints, and the values for those keys are the disks that will
+// be returned from DiskFromMountPoint or used internally in
+// MountPointIsFromDisk.
+func MockMountPointDisksToPartionMapping(mockedMountPoints map[Mountpoint]MockDiskMapping) (restore func()) {
+	osutil.MustBeTestBinary("mock disks only to be used in tests")
+
+	old := diskFromMountPoint
+
```

Comment:
this maybe should check that all the MockDiskMapping have distrinct DevNums ?

Reply:
no because that's not always going to be actually the case actually, if you look at the usage from the mega PR, we have instances like this for run mode:

```go
	restore := disks.MockMountPointDisksToPartionMapping(
		map[disks.Mountpoint]disks.MockDiskMapping{
			{Mountpoint: boot.InitramfsUbuntuBootDir}: defaultBootDisk,
			{Mountpoint: boot.InitramfsDataDir}:       defaultBootDisk,
		},
	)
	defer restore()
```

where different mount points go to the same disk

Reply:
mmh, maybe a more realistic mapping would be:

MountPoint -> filesystemlabel map + list of MockDiskMappings

Reply:
sorry,  let me step back a bit I think the thing that is tripping me here is:

map[Mountpoint]MockDiskMapping

we are using MockDiskMapping as a value but it has a map inside, and really a disk exists once in the world, and that structure is not quite what I would call just a labeling for a disk.  Now if we do this:

map[Mountpoint]*MockDiskMapping

we can make a map  of distinct MockDiskMappings map[*MockDiskMapping]true  and over the keys of that DevNum should not repeat.

Reply:
yes we worked this out over IRC, hopefully the new pattern is what we want